FLT/DedekindDomain/FiniteAdeleRing/IsDirectLimitRestricted.lean

```diff
@@ -0,0 +1,182 @@
+  left_inv x := by
+    simp only [dite_eq_ite, ite_self]
+    rfl
+  right_inv x := by
+    simp only [mk_apply, Subtype.coe_prop, ↓reduceDIte, Subtype.coe_eta]
+    ext i
+    · rfl
+    · simp only [dif_neg i.property]
+
+end type
+
+variable {T : ι → Type*} [Π i, SetLike (T i) (R i)] {A : Π i, T i}
+
+section monoid
+
+/-- Monoid equivalence version of `principalEquivProd`. -/
+@[to_additive /-- Additive monoid equivalence of principalEquivProd. -/]
+def principalMulEquivProd [Π i, Monoid (R i)] [∀ i, SubmonoidClass (T i) (R i)] :
+    Πʳ i, [R i, A i]_[𝓟 S] ≃* (Π i : {i // i ∈ S}, A i) × (Π i : {i // i ∉ S}, R i) where
+  __ := principalEquivProd R S _
+  map_mul' _ _ := rfl
+
+end monoid
+
+variable {ι : Type*} (R : ι → Type*) {ℱ : Filter ι} (A : Type*) [CommRing A]
+
+open scoped RestrictedProduct TensorProduct
+
+open Filter
+
+section module
+
+/-- Module equivalence version of `principalEquivProd`. -/
+noncomputable def principal [Π i, AddCommGroup (R i)]
+    [∀ i, Module A (R i)] {C : ∀ i, Submodule A (R i)}
+    (S : Set ι) [∀ i, Decidable (i ∈ S)] :
+   (Πʳ i, [R i, C i]_[𝓟 S]) ≃ₗ[A] ((Π i : {i // i ∈ S}, C i) ×
+  (Π i : {i // i ∉ S}, R i)) where
+    __ := principalAddEquivSum R S (A := C)
+    map_smul' m x := by
+      simp only [AddEquiv.toEquiv_eq_coe, Equiv.toFun_as_coe, EquivLike.coe_coe, RingHom.id_apply]
+      dsimp [principalAddEquivSum, principalEquivProd]
+      ext i
+      · simp only [Pi.smul_apply, SetLike.coe_sort_coe, SetLike.val_smul]
+      · simp only [Pi.smul_apply]
+
+variable [Π i, AddCommGroup (R i)] [∀ i, Module A (R i)] {C : ∀ i, Submodule A (R i)}
+
+/-- If `𝓕 ≤ 𝓖`, the restricted product `Πʳ i, [R i, A i]_[𝓖]` is naturally included in
+`Πʳ i, [R i, A i]_[𝓕]`. This is the corresponding map. -/
+def linclusion
+    {ι : Type*} {R₀ : Type*} (R : ι → Type*) [Semiring R₀] [∀ i, AddCommMonoid (R i)]
+    [∀ i, Module R₀ (R i)] (A : (i : ι) → Submodule R₀ (R i)) {ℱ 𝓖 : Filter ι}
+    (h : ℱ ≤ 𝓖) : Πʳ i, [R i, A i]_[𝓖] →ₗ[R₀] Πʳ i, [R i, A i]_[ℱ] where
+  toFun := inclusion R (A ·) h
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+instance {I : Type*} [Preorder I] (𝓖 : I → Filter ι) (h𝓖 : Antitone 𝓖) :
+    DirectedSystem (fun x ↦ Πʳ (i : ι), [R i, ↑(C i)]_[𝓖 x]) (linclusion _ _ <| @h𝓖 · · ·) where
+  map_self _ _ := rfl
+  map_map _ _ _ _ _ _ := rfl
+
+instance instIsDirectLimit {I : Type*} [Preorder I] [Nonempty I] [IsDirected I (· ≤ ·)]
+    (𝓖 : I → Filter ι) (h𝓖 : Antitone 𝓖) (hℱ : ℱ = iInf 𝓖) :
+    IsDirectLimit (Πʳ i, [R i, C i]_[𝓖 ·]) Πʳ i, [R i, C i]_[ℱ]
+    (linclusion _ _ <| @h𝓖 · · ·) (linclusion _ _ <| hℱ.trans_le <| iInf_le 𝓖 ·) where
+  inj Sᵢ Sⱼ mi mj hmij := by
+    obtain ⟨Sₖ, hik, hjk⟩ := @directed_of _ (· ≤ ·) _ Sᵢ Sⱼ
+    refine ⟨Sₖ, hik, hjk, ?_⟩
+    dsimp [RestrictedProduct.inclusion, RestrictedProduct.inclusion]
+    dsimp [coeLinearMap,
+      RestrictedProduct.inclusion] at hmij
+    apply Subtype.ext
+    simp only
+    injection hmij
```

Comment:
```suggestion
    refine ⟨Sₖ, hik, hjk, ?_⟩
    apply Subtype.ext
    injection hmij
```
These calls to the simplifier are only changing things up to definitional equality and are not needed. Same for the surj and compatibility proofs.